src/Mapping/ExtensionMetadataFactory.php

```diff
@@ -66,8 +69,15 @@ class ExtensionMetadataFactory
      */
     private $cacheItemPool;
 
+    /**
+     * @param Reader|AttributeReader $annotationReader
+     */
     public function __construct(ObjectManager $objectManager, string $extensionNamespace, object $annotationReader, ?CacheItemPoolInterface $cacheItemPool = null)
     {
+        if (!$annotationReader instanceof Reader && !$annotationReader instanceof AttributeReader) {
+            trigger_deprecation('gedmo/doctrine-extensions', '3.11', 'Providing an annotation reader which does not implement %s or is an instance of %s is deprecated, as of 4.0 the reader must implement the interface.', Reader::class, AttributeReader::class);
```

Comment:
Please, try to use a message that includes the reference to the received object type in order to help the users debug the problem.